patches/server/0965-Make-banners-modifiable-with-both-BannerMeta-and-Blo.patch

```diff
@@ -0,0 +1,270 @@
++        super(meta, material);
+ 
+-        if (!(meta instanceof CraftMetaBanner)) {
++        if (true || !(meta instanceof CraftMetaBanner)) {
++            this.bannerBlock = (org.bukkit.block.Banner) this.getBlockState();
++            // Paper end - Create new constructor to properly handle different banners
+             return;
+         }
+ 
+@@ -74,11 +89,20 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
+         this.base = banner.base;
+         this.patterns = new ArrayList<Pattern>(banner.patterns);
+     }
+-
++    // Paper start - Create new constructor to properly handle different banners
++    @Deprecated
+     CraftMetaBanner(CompoundTag tag) {
+-        super(tag);
++        this(tag, Material.WHITE_BANNER);
++
++        new org.bukkit.plugin.AuthorNagException("Usage of internal deprecated constructor in paper: CraftMetaBanner(CompoundTag), this will incorrectly assume the item is a white banner").printStackTrace();
++    }
+ 
+-        if (!tag.contains("BlockEntityTag")) {
++    CraftMetaBanner(CompoundTag tag, Material material) {
++        super(tag, material);
++
++        if (true || !tag.contains("BlockEntityTag")) {
++            this.bannerBlock = (org.bukkit.block.Banner) this.getBlockState();
++            // Paper end - Create new constructor to properly handle different banners
+             return;
+         }
+ 
+@@ -102,6 +126,12 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
+ 
+     CraftMetaBanner(Map<String, Object> map) {
+         super(map);
++        // Paper start - Use underlying banner block state
++        if (true) {
++            this.bannerBlock = (org.bukkit.block.Banner) this.getBlockState();
++            return;
++        }
++        // Paper end - Use underlying banner block state
+ 
+         String baseStr = SerializableMeta.getString(map, BASE.BUKKIT, true);
+         if (baseStr != null) {
+@@ -123,6 +153,7 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
+     @Override
+     void applyToItem(CompoundTag tag) {
+         super.applyToItem(tag);
++        if (true) return; // Paper - CraftMetaBlockState already save the banner for us
+ 
+         CompoundTag entityTag = new CompoundTag();
+         if (this.base != null) {
+@@ -144,46 +175,85 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
+ 
+     @Override
+     public DyeColor getBaseColor() {
++        if (true) return this.bannerBlock.getBaseColor(); // Paper - Use underlying block state
+         return this.base;
+     }
+ 
+     @Override
+     public void setBaseColor(DyeColor color) {
++        // Paper start - Use underlying block state
++        if (true) {
++            this.bannerBlock.setBaseColor(color);
++            this.setBlockState(this.bannerBlock); // Send changes to the underlying class
++            return;
++        }
++        // Paper end - Use underlying block state
```

Comment:
This should get the blockstate, make the change, and then call `this.setBlockState()`.
You don't want to have an instance of the blockstate that could easily become desynced with what the underlying `CompoundTag` in `CraftMetaBlockState` shows. It's easier to just create and set as needed. Yes, its less efficient, but we are deprecating it partly because of that.